src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
+    :raises InternalError: if a kernel argument has a shape defined by \
+        something other than ArrayType.Extent or a Reference.
+
+    '''
+    symbol_table = SymbolTable()
+
+    # TODO #1331 provide some way of configuring the extent of the test arrays
+    array_dim_size = 20
+
+    # We expect a single Container containing a single Kernel. Anything else
+    # is not supported. However, we have to allow for the fact that, in
+    # general, there will be an outermost FileContainer.
+    container = _find_container(ad_psyir)
+    if not container:
+        raise NotImplementedError(
+            "Generation of a test harness is only supported for a TL kernel "
+            "implemented as a subroutine within a module but failed to find "
+            "enclosing module.")
+
+    # First check that there's only one routine and that it's not a Program.
+    routines = tl_psyir.walk(Routine)
+
+    if len(routines) != 1:
+        raise NotImplementedError(
+            "The supplied Fortran must contain one and only one subroutine "
+            "but found: {0}".format([sub.name for sub in routines]))
+
+    tl_kernel = routines[0]
+
+    if tl_kernel.is_program:
+        raise NotImplementedError(
+            "Generation of a test harness for a kernel defined as a Program "
+            "(as opposed to a Subroutine) is not currently supported. (Found "
+            "'{0}' which is a Program.)".format(tl_kernel.name))
+
+    # First Container is a FileContainer and that's not what we want
+    container = tl_psyir.walk(Container)[1]
+
+    # Get the Container and Routine names from the PSyIR of the adjoint.
+    adjoint_kernel_name = ad_psyir.walk(Routine)[0].name
+    adjoint_module_name = ad_psyir.walk(Container)[1].name
+
+    # Create a symbol for the TL kernel
+    csym = ContainerSymbol(container.name)
+    symbol_table.add(csym)
+    tl_kernel_sym = tl_kernel.symbol_table.lookup(tl_kernel.name).copy()
+    tl_kernel_sym.interface = GlobalInterface(csym)
+    symbol_table.add(tl_kernel_sym)
+
+    # Create a symbol for the adjoint kernel
+    adj_container = ContainerSymbol(adjoint_module_name)
+    symbol_table.add(adj_container)
+    adj_kernel_sym = symbol_table.new_symbol(
+        adjoint_kernel_name, symbol_type=RoutineSymbol,
+        interface=GlobalInterface(adj_container))
+
+    # Create a symbol to hold the extent of any test arrays. This is done here
+    # to avoid any clashes with any of the container and kernel names.
+    dim_size_sym = symbol_table.new_symbol("array_extent",
+                                           symbol_type=DataSymbol,
+                                           datatype=INTEGER_TYPE,
+                                           constant_value=array_dim_size)
+
+    # Create symbols for the results of the inner products
+    inner1 = symbol_table.new_symbol("inner1", symbol_type=DataSymbol,
+                                     datatype=REAL_DOUBLE_TYPE)
```

Comment:
At some point we will need to deal with specifying precision. Might be worth adding a new issue and reference it here. I'm thinking that we will need some sort of config (which could just be a class, or could actually be a config file) and/or command line switches to specify what we want.

Reply:
I'm not sure about that. Since this is to store the result of the inner product, aren't we always going to want that to be as precise as possible, irrespective of the precision used by a kernel?

Reply:
What you say makes sense but I don't know what they expect. It's probably a good thing to raise at the next telco, or ask on teams. However, it feels wrong (to me) to hard code a particular precision.

Reply:
Yep, done (and question asked).